native/src/main/java/io/ballerina/stdlib/data/xml/XmlTraversal.java

```diff
@@ -0,0 +1,532 @@
+import io.ballerina.runtime.api.types.Field;
+import io.ballerina.runtime.api.types.MapType;
+import io.ballerina.runtime.api.types.RecordType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.XmlNodeType;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.utils.TypeUtils;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BXml;
+import io.ballerina.runtime.api.values.BXmlItem;
+import io.ballerina.runtime.api.values.BXmlSequence;
+import io.ballerina.stdlib.data.utils.Constants;
+import io.ballerina.stdlib.data.utils.DataUtils;
+import io.ballerina.stdlib.data.utils.DataUtils.XmlAnalyzerData;
+import io.ballerina.stdlib.data.utils.DiagnosticErrorCode;
+import io.ballerina.stdlib.data.utils.DiagnosticLog;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.xml.namespace.QName;
+
+/**
+ * Convert Xml value to a ballerina record.
+ *
+ * @since 0.1.0
+ */
+
+public class XmlTraversal {
+
+    private static final ThreadLocal<XmlTree> tlXmlTree = ThreadLocal.withInitial(XmlTree::new);
+
+    public static Object traverse(BXml xml, Type type) {
+        XmlTree xmlTree = tlXmlTree.get();
+        return xmlTree.traverseXml(xml, type);
+    }
+
+    static class XmlTree {
+        private Object currentNode;
+
+        @SuppressWarnings("unchecked")
+        public Object traverseXml(BXml xml, Type type) {
+            Type referredType = TypeUtils.getReferredType(type);
+            switch (referredType.getTag()) {
+                case TypeTags.RECORD_TYPE_TAG:
+                    XmlAnalyzerData analyzerData = new XmlAnalyzerData();
+                    RecordType recordType = (RecordType) referredType;
+                    currentNode = ValueCreator.createRecordValue(recordType);
+                    BXml nextXml = validateRootElement(xml, recordType, analyzerData);
+                    Object resultRecordValue = traverseXml(nextXml, recordType, analyzerData);
+                    DataUtils.validateRequiredFields((BMap<BString, Object>) resultRecordValue, analyzerData);
+                    return resultRecordValue;
+                case TypeTags.MAP_TAG:
+                    MapType mapType = (MapType) referredType;
+                    RecordType anonRecType = TypeCreator.createRecordType(Constants.ANON_TYPE, mapType.getPackage(), 0,
+                            new HashMap<>(), mapType.getConstrainedType(), false, 0);
+                    return traverseXml(xml, anonRecType);
+                default:
+                    return DiagnosticLog.error(DiagnosticErrorCode.UNSUPPORTED_TYPE, Constants.RECORD_OR_MAP,
+                            type.getName());
+            }
+        }
```

Comment:
In `XMLParser.java`, you have used if-else-if blocks instead of a switch statement for similar case as this. It would be better to have a consistent approach, and IMO, the switch statement looks better.